changelogs/unreleased/persistent_gc_state.md

```diff
@@ -1,4 +1,5 @@
 ## feature/replication
 
```

Comment:
> What is the motivation behind making the space read only on recovery? It looks quite weird, that we write there and take the state immediately, but any other write is accepted and yet makes no effect until restart. Could as well disallow touching this space completely by users? Make it read-only for them? Otherwise it is extremely confusing now.

Users shouldn't touch system spaces anyway. Still, there should be a way to update them directly in case of a disaster at the user's own risk so I don't think we should explicitly forbid that in the code.

> Tuples of the same space, even all being valid and up to date and not behind the latest snapshot, can be working in 3 different ways. That looks like a sure way to get lots of confusion from users and bugs. Tbh, what is even a bug if we consider this state okay?
> 
> Why is it so complex just to handle updates in `on_commit` trigger of `_gc_consumers`?

`_gc_consumers` is different from other system spaces by design. While other spaces must be kept in sync with the in-memory data dictionary state, `_gc_consumers` isn't a source of truth. The actual gc state is stored in memory, as before, and is periodically written to `_gc_consumers` in the background. Tarantool reads it only on recovery in order to restore the in-memory gc state. IMO such a design is simpler than strict syncing with the aid of replace/commit triggers (the code is almost completely isolated in `gc.c`). The primary reason why we use replace/commit triggers for other system spaces is that this lets us handle replication, recovery, and normal operation in the same way. For gc there's no need for that: the state is local to the instance, updated in the background, and may lag.

Reply:
> Users shouldn't touch system spaces anyway.

I don't mean in a sense that they should touch it, sorry, I should have been more specific. And you are right, they shouldn't touch them (in regular work). I rather mean that it is hard to explain how this works.

As you can see, the GC consumer tuples can be in 3 different states. Intuitively when I look at a space, I expect it to be reflecting the actual single state, doesn't matter if I can change it or not. But here even with documentation I am going to have a hard time to get how this works if I am a user and, say, I am debugging why my xlogs are being deleted. I did really have some troubles even as a tnt developer to wrap my mind around it.

Really, imagine user's PoV. They see that xlogs are being deleted. But they know that the GC consumers are persistent! Saw it in the changelog or something. So they get the list of spaces, find there `_gc_consumers`, hm, seems like the place. Then they see - aha, some replica is missing or is outdated or whatever. So they insert/update a tuple aaaand ... nothing happens. Stuff still gets deleted. That looks to me like a likely and valid user story.

Or vice versa - they see something isn't being deleted. And they need disk space. So they delete a tuple. A dead anon replica's GC consumer. But again nothing happens? Does it also mean, that if I got a new anon consumer created (not having a tuple in `_cluster`), and then I want to delete it, then I can only do it manually, and manually it won't do anything until restart? Thus to unblock deletion of xlogs after an anon replica's death I need to restart the master. Or did I miss something?

> _gc_consumers is different from other system spaces by design

Different in design and how it works, but not from external and internal user's PoV, for whom this space looks like a regular space with a regular engine. They will probably see this space and wonder why doesn't it work as they expect intuitively. I would bet money that if we ask in the "Tarantool Team" chat among our internal users how do they think `_gc_consumers` works, nobody would guess correctly how it works. It is quite counter-intuitive ATM. Whether the behaviour is the problem, or the name of the space suggesting that this is the actual state - can't tell myself.

Reply:
Allowing users to manage in-memory state through the space will significantly complicate the patch, and, at the first glance, it will require complete re-design of the feature.

We have special doc sections for system spaces (for example, [`_cluster`](https://www.tarantool.io/ru/doc/latest/reference/reference_lua/box_space/_cluster/)). We can describe the behavior there, something like that:
```
The space is read right after recovery and then Tarantool only dumps its in-memory state there from time to time. Hence, any manual modification of this space will take an effect only after restart.
Do not use this space for WAL GC analysis - use `box.info.gc().consumers` instead.
```

What's about persistent consumers of anon replicas: we will think about them in another patch =)
Right now, if replica registers consumer with `IPROTO_REGISTER`, it can unregister it only with `IPROTO_REGISTER` with `IS_PERSISTENT_GC = false`. Admin now cannot manually delete consumer of anonymous replica, but we can introduce `box.ctl.delete_gc_consumer` that will call `gc_erase_consumer` under the hood.